tools.py

```diff
@@ -6,13 +7,17 @@
 import itertools
 import subprocess
 import collections
+from functools import partial
```

Comment:
Please keep the import style consistent with the existing code, i.e. in this codebase I prefer to always deal with module names at callsite (especially if calls are few like in this case):
```
import functools

functools.partial(...)
```

Reply:
fixed